docs/source/python/compute.rst

```diff
@@ -370,3 +370,133 @@ our ``even_filter`` with a ``pc.field("nums") > 5`` filter:
 
 :class:`.Dataset` currently can be filtered using :meth:`.Dataset.to_table` method
 passing a ``filter`` argument. See :ref:`py-filter-dataset` in Dataset documentation.
+
+
+User-Defined Functions
+======================
+
+.. warning::
+   This API is **experimental**.
+
+PyArrow allows defining and registering custom compute functions.
+These functions can then be called from Python as well as C++ (and potentially
+any other implementation wrapping Arrow C++, such as the R ``arrow`` package)
+using their registered function name.
+
+To register a UDF, a function name, function docs, input types and
+output type need to be defined. Using :func:`pyarrow.compute.register_scalar_function`,
+
+.. code-block:: python
+
+   import numpy as np
+
+   import pyarrow as pa
+   import pyarrow.compute as pc
+
+   function_name = "numpy_gcd"
+   function_docs = {
+         "summary": "Calculates the greatest common divisor",
+         "description":
+            "Given 'x' and 'y' find the greatest number that divides\n"
+            "evenly into both x and y."
+   }
+
+   input_types = {
+      "x" : pa.int64(),
+      "y" : pa.int64()
+   }
+
+   output_type = pa.int64()
+
+   def to_np(val):
+      if isinstance(val, pa.Scalar):
+         return val.as_py()
```

Comment:
Is this conversion to Scalar still needed? (with the latest refactor of Wes, scalars might now be handled by len-1 arrays?) 
In any case, the `gcd_numpy` function itself won't work with scalars because of the `pa.array(..)` call in it:

```
In [32]: gcd_numpy(None, pa.scalar(27), pa.scalar(63))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-32-5dc8dd5d05b1> in <module>
----> 1 gcd_numpy(None, pa.scalar(27), pa.scalar(63))

<ipython-input-26-1579a8ef575a> in gcd_numpy(ctx, x, y)
     22    np_x = to_np(x)
     23    np_y = to_np(y)
---> 24    return pa.array(np.gcd(np_x, np_y))
     25 pc.register_scalar_function(gcd_numpy,
     26                            function_name,

~/scipy/repos/arrow/python/pyarrow/array.pxi in pyarrow.lib.array()

~/scipy/repos/arrow/python/pyarrow/array.pxi in pyarrow.lib._sequence_to_array()

~/scipy/repos/arrow/python/pyarrow/error.pxi in pyarrow.lib.pyarrow_internal_check_status()

TypeError: 'numpy.int64' object is not iterable
```

Reply:
@jorisvandenbossche if we do that, we get the following error

```bash
>>> pc.call_function("numpy_gcd", [pa.scalar(27), pa.array([81, 12, 5])])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/_compute.pyx", line 560, in pyarrow._compute.call_function
  File "pyarrow/_compute.pyx", line 355, in pyarrow._compute.Function.call
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/_compute.pyx", line 2506, in pyarrow._compute._scalar_udf_callback
  File "<stdin>", line 4, in gcd_numpy
  File "/Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/numpy/core/_internal.py", line 790, in _gcd
    a, b = b, a % b
TypeError: unsupported operand type(s) for %: 'pyarrow.lib.Int64Scalar' and 'int'
```

I think the reason is, Numpy cannot identify the passed in Arrow scalar value. We need to take the python value of it or convert it to numpy. 

The following is what would happen 

```bash
>>> np.gcd(np.array(pa.scalar(27)), np.array([81, 12, 5]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/numpy/core/_internal.py", line 790, in _gcd
    a, b = b, a % b
TypeError: unsupported operand type(s) for %: 'pyarrow.lib.Int64Scalar' and 'int'
```

But it would work for 

```bash
np.gcd(np.array(pa.array([27])), np.array([81, 12, 5]))
array([27,  3,  1])
```

but not for 

```bash
>>> np.gcd(np.array(pa.scalar(27)), np.array([81, 12, 5]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/numpy/core/_internal.py", line 790, in _gcd
    a, b = b, a % b
TypeError: unsupported operand type(s) for %: 'pyarrow.lib.Int64Scalar' and 'int'
```

And again works for


```bash
>>> np.gcd(np.array(27), np.array([81, 12, 5]))
array([27,  3,  1])
```

Am I wrong here or missing something?

Reply:
OK, so it seems that we are _still_ passing scalars to the UDF in case of mixed scalar / array arguments, like you do in `pc.call_function("numpy_gcd", [pa.scalar(27), pa.array([81, 12, 5])])`. 

So my understanding was wrong based on testing the case of two scalars: `pc.call_function("numpy_gcd", [pa.scalar(27), pa.scalar(81)])`. That actually works by passing two length-1 arrays to the UDF. 
(and for that reason I assumed we always pass scalars as length-1 arrays to the UDF implementation)

Personally, I find this a bit confusing (but I am not familiar enough with the internals of the kernels to know if it is easy to always pass arrays to the UDF). And at least I would document that the arguments can be scalars, but never all of them will be scalar at the same time. 
(since as I mentioned above, the current example UDF implementation doesn't work if it would be passed only scalars)

Reply:
I understand your point. That's why the description included that idea. We didn't change the UDF kernels since it was merged. So for the moment I will document what's missing from the current docs. cc @westonpace @pitrou 
What's the best?